Ubuntu/History/historymanager.cpp

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (C) 2021 Ubports Foundation
+ *
+ * Authors:
+ *  Lionel Duboeuf <lduboeuf@ouvaton.org>
+ *
+ * This file is part of history-service.
+ *
+ * history-service is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * history-service is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "historymanager.h"
+#include "manager.h"
+#include "eventview.h"
+#include <QDebug>
+
+
+HistoryManager::HistoryManager(QObject *parent) : QObject(parent), mFilter(0), mSort(new HistoryQmlSort(this)), mType(HistoryModel::EventTypeText), mDeletedCount(0), mPendingOperation(false), mSignalsTimer(-1)
+{
+}
+
+HistoryQmlFilter *HistoryManager::filter() const
+{
+    return mFilter;
+}
+
+void HistoryManager::setFilter(HistoryQmlFilter *value)
+{
+    if (mFilter) {
+        mFilter->disconnect(this);
+    }
+
+    mFilter = value;
+
+    Q_EMIT filterChanged();
+}
+
+HistoryModel::EventType HistoryManager::type() const
+{
+    return mType;
+}
+
+void HistoryManager::setType(HistoryModel::EventType value)
+{
+    mType = value;
+    Q_EMIT typeChanged();
+}
+
+HistoryQmlSort *HistoryManager::sort() const
+{
+    return mSort;
+}
+
+int HistoryManager::count() const
+{
+    return mCount;
+}
+
+int HistoryManager::deletedCount() const
+{
+    return mDeletedCount;
+}
+
+void HistoryManager::setSort(HistoryQmlSort *value)
+{
+    // disconnect the previous sort
+    if (mSort) {
+        mSort->disconnect(this);
+    }
+
+    mSort = value;
+
+    Q_EMIT sortChanged();
+}
+
+int HistoryManager::eventsCount()
+{
+    History::Filter queryFilter;
+
+    if (!mType || !mFilter) {
+        return false;
+    }
+
+    if (mFilter && mFilter->filter().isValid()) {
+        queryFilter = mFilter->filter();
+    }
+
+    return History::Manager::instance()->eventsCount((History::EventType)mType, queryFilter);
+}
+
+void HistoryManager::timerEvent(QTimerEvent *event)
+{
+    if (event->timerId() == mSignalsTimer) {
+        pendingOperation(false);
+    }
+}
+
+bool HistoryManager::removeAll()
+{
+
+    History::Filter queryFilter;
+    History::Sort querySort;
+
+    if (mPendingOperation) {
+        qWarning() << "there is a pending operation, request cancelled";
+        return false;
+    }
+
+    if (!mType || !mFilter || !mFilter->filter().isValid()) {
+        qWarning() << "type or filter not set, operation cancelled";
+        return false;
```

Comment:
This doesn't seem tenable. A client could ask for history deletion but end up receiving `false`. Now it doesn't know if it should correct its parameters and retry or just retry later after the pending operation is finished. Exceptions, specific return values, or some other solution is needed.